backends/gstreamer/src/player.rs

```diff
@@ -17,16 +18,19 @@ use std::time;
 use std::u64;
 
 fn frame_from_sample(sample: &gst::Sample) -> Result<Frame, ()> {
-    let caps = sample.get_caps().ok_or_else(|| ())?;
-    let s = caps.get_structure(0).ok_or_else(|| ())?;
-    let width = s.get("width").ok_or_else(|| ())?;
-    let height = s.get("height").ok_or_else(|| ())?;
-
     let buffer = sample.get_buffer().ok_or_else(|| ())?;
-    let map = buffer.map_readable().ok_or_else(|| ())?;
-    let data = Vec::from(map.as_slice());
-
-    Ok(Frame::new(width, height, Arc::new(data)))
+    let info = sample
+        .get_caps()
+        .and_then(|caps| VideoInfo::from_caps(caps.as_ref()))
+        .ok_or_else(|| ())?;
+    let frame = VideoFrame::from_buffer_readable(buffer, &info).or_else(|_| Err(()))?;
+    let data = frame.plane_data(0).ok_or_else(|| ())?;
+
+    Ok(Frame::new(
+        info.width() as i32,
+        info.height() as i32,
+        Arc::new(data.to_vec()),
```

Comment:
This still copies like it did before by requiring to have a `Vec<u8>` in the API. Ideally you would change the API so that it takes an `Arc<AsRef<[u8]>>` instead: in that case you could simply pass the return value of `buffer.map_readable()` in there.

My old servo branch had it implemented without copying similar to that.